### ISSUE TITLE
Failed put operations do not result in error

### ISSUE THREAD

**Issue by warwickmm** [duplicati/duplicati#3673]
If all the `Put` operations for a backup fail, no error is reported to the user.  Even worse, the UI updates the "Last successful backup" time, which can lead the user to believe that the backup was successful.  The backup log shows a green check mark, and indicates 0 errors and 0 warnings.  

![image](https://user-images.githubusercontent.com/17345532/53294374-6d509680-379a-11e9-8f43-81a856c282d6.png)

![image](https://user-images.githubusercontent.com/17345532/53294430-670eea00-379b-11e9-9116-adb291bb8276.png)

If this occurs on the first backup, the user might sense something is wrong if they happen to see "Backup:0 bytes / 0 Versions".  If this occurs on a subsequent backup, it's unclear if there's any way for the user to know that the backup was unsuccessful.

Sample logs here:
```
  "WarningsActualLength": 0,
  "ErrorsActualLength": 0,
  "Messages": [
    "2019-02-23 18:33:36 -08 - [Information-Duplicati.Library.Main.Controller-StartingOperation]: The operation Backup has started",
    "2019-02-23 18:33:36 -08 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: List - Started:  ()",
    "2019-02-23 18:33:36 -08 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: List - Completed:  ()",
    "2019-02-23 18:33:36 -08 - [Information-Duplicati.Library.Main.Operation.FilelistProcessor-SchedulingMissingFileForDelete]: scheduling missing file for deletion, currently listed as Uploading: duplicati-20190224T023001Z.dlist.zip.aes",
    "2019-02-23 18:33:36 -08 - [Information-Duplicati.Library.Main.Operation.FilelistProcessor-RemoteUnwantedMissingFile]: removing file listed as Temporary: duplicati-bc4d8df45af10459f8587af1a0f257f8f.dblock.zip.aes",
    "2019-02-23 18:33:36 -08 - [Information-Duplicati.Library.Main.Operation.FilelistProcessor-RemoteUnwantedMissingFile]: removing file listed as Deleting: duplicati-be48686392a7c4b0ca436bbc410529767.dblock.zip.aes",
    "2019-02-23 18:33:36 -08 - [Information-Duplicati.Library.Main.Operation.FilelistProcessor-RemoteUnwantedMissingFile]: removing file listed as Deleting: duplicati-b640d1ceda8964cb58ff6490b9e8a69ad.dblock.zip.aes",
    "2019-02-23 18:33:36 -08 - [Information-Duplicati.Library.Main.Operation.FilelistProcessor-RemoteUnwantedMissingFile]: removing file listed as Deleting: duplicati-ba676d62a43914d90bd5d225c41226fee.dblock.zip.aes",
    "2019-02-23 18:33:36 -08 - [Information-Duplicati.Library.Main.Operation.FilelistProcessor-RemoteUnwantedMissingFile]: removing file listed as Deleting: duplicati-bd74b19ede236418a85a32b7443b30d80.dblock.zip.aes",
    "2019-02-23 18:33:40 -08 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Started: duplicati-bb3dd236e8e2f4803a3cc0305e98f8f24.dblock.zip.aes (11.03 MB)",
    "2019-02-23 18:33:40 -08 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Retrying: duplicati-bb3dd236e8e2f4803a3cc0305e98f8f24.dblock.zip.aes (11.03 MB)",
    "2019-02-23 18:33:50 -08 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Rename: duplicati-bb3dd236e8e2f4803a3cc0305e98f8f24.dblock.zip.aes (11.03 MB)",
    "2019-02-23 18:33:50 -08 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Rename: duplicati-b1e56b463e16640ddaa4778ce0542357b.dblock.zip.aes (11.03 MB)",
    "2019-02-23 18:33:50 -08 - [Information-Duplicati.Library.Main.Operation.Common.BackendHandler-RenameRemoteTargetFile]: Renaming \"duplicati-bb3dd236e8e2f4803a3cc0305e98f8f24.dblock.zip.aes\" to \"duplicati-b1e56b463e16640ddaa4778ce0542357b.dblock.zip.aes\"",
    "2019-02-23 18:33:50 -08 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Started: duplicati-b1e56b463e16640ddaa4778ce0542357b.dblock.zip.aes (11.03 MB)",
    "2019-02-23 18:33:50 -08 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Retrying: duplicati-b1e56b463e16640ddaa4778ce0542357b.dblock.zip.aes (11.03 MB)",
    "2019-02-23 18:34:00 -08 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Rename: duplicati-b1e56b463e16640ddaa4778ce0542357b.dblock.zip.aes (11.03 MB)",
    "2019-02-23 18:34:00 -08 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Rename: duplicati-b997c64a4e92a4987bde9b1a19e610b9e.dblock.zip.aes (11.03 MB)",
    "2019-02-23 18:34:00 -08 - [Information-Duplicati.Library.Main.Operation.Common.BackendHandler-RenameRemoteTargetFile]: Renaming \"duplicati-b1e56b463e16640ddaa4778ce0542357b.dblock.zip.aes\" to \"duplicati-b997c64a4e92a4987bde9b1a19e610b9e.dblock.zip.aes\"",
    "2019-02-23 18:34:00 -08 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Started: duplicati-b997c64a4e92a4987bde9b1a19e610b9e.dblock.zip.aes (11.03 MB)"
  ],
  "Warnings": [],
  "Errors": [],
```
## Environment info

- **Duplicati version**: 2.0.4.15
- **Operating system**: All
- **Backend**: All

## Steps to reproduce
1. Configure a backup to a local destination.
2. Make sure that the local destination folder is not writable.
3. Run the backup and observe in the live Profile logs that the Put operations fail.
4. No error is reported to the user, and the UI appears to indicate that the backup was successful.  No files appears in the destination folder.

**Comment by warwickmm** [comment]
I'm still unfamiliar with all the async stuff, but I suspect that something is going on [here](https://github.com/duplicati/duplicati/blob/master/Duplicati/Library/Main/Operation/BackupHandler.cs#L359).  In the example above, the `uploader` `Task` throws an `Exception` and does not run to completion.  The call to `Task.WhenAny` returns the `flushReq.LastWriteSizeAync` `Task`, the `uploader` `Task` is not awaited, and hence the `Exception` is never raised. 

@kenkendk, do you have any additional insight here?  This seems like a critical issue.

**Comment by ts678** [comment]
After the dblock fails (see log), does live log show Duplicati continue on to dindex or dlist put? Maybe this problem is related to [Error / DB corruption: Unexpected difference in fileset version X: found Y entries, but expected Z #3644](https://github.com/duplicati/duplicati/issues/3644#issuecomment-463306796) which can be reproduced in several ways (I'm linking to mine which is silent corruption, success reported, and damage left for next backup -- how healthy is your next backup, after put is fixed?).

**Comment by warwickmm** [comment]
After the `dblock` file, Duplicati continues on to try and put the `dlist`file, which also fails (as expected).  I'm not sure if that issue is related.  In my test case, if I later make the destination folder writable, then backups resume without any issue and restores work just fine.

**Comment by ts678** [comment]
From a cause and sequence view they're similar. If one follows the links to read my [testing summary](https://forum.duplicati.com/t/unexpected-difference-in-fileset-errors-2-0-3-9/4286/19?u=ts678), we both had a dblock put failure, a dindex not tried, and a dlist put (except yours was blocked, and mine was allowed). If yours had been allowed, it might have wound up just like mine. Everything else seems the same, doesn't it?

>  I'm not sure if that issue is related.

I'm not sure either, but if you'd like to go a step further, we can find out. As a side note, my test case initially was attempted with Linux because a chmod is so easy to do, but it didn't cause the corruption (for unknown reasons) so I went back to try a network failure but at least used an easier-to-get server than what originally stumbled over the problem. If you'd rather not try my test case, maybe I'll try it using Windows permissions.

**Comment by warwickmm** [comment]
I think I was able to replicate your test case.  I ran a backup, shut down my network connection while attempting to put `dblock` files, and brought up my connection while attempting to put `dlist` files.  The backup reported no errors, but the subsequent backup resulted in an error.

Perhaps if the failure to put the `dblock` files halted the backup, the database would hopefully not be corrupted by the errant entries.

**Comment by seantempleton** [comment]
I was looking around the issues list to see what was out here and stumbled across this. On my Windows machine I set a folder to not be writable and was able to duplicate the issue.

I then tried it using my code changes for parallel uploads and the backup failed immediately once it started to upload. If I can get my code into a state that it is accepted and then merged this might be resolved then.

I think the problem is [this line](https://github.com/duplicati/duplicati/blob/c3136e31dfcf3c0857ed7f5eecf6cfb362d56df7/Duplicati/Library/Main/Operation/Backup/BackendUploader.cs#L148). From the documentation `Task.WhenAny` will have as its result the first task that completed, even if that task was in the canceled or faulted state. If the task doing the upload goes into the faulted state, then when `WhenAny` is executed on that line the result from `WhenAny` will equal `top.Value`. Execution will not go in to the if statement and call await and hence no exception will be raised.

**Comment by warwickmm** [comment]
Thanks @seantempleton for looking into this.  I had the same thought with the changes you're making, but unfortunately I still do not encounter an error (using revision 22369abbdedab758993eb7682e9c50139f24656e in your branch).  In my brief testing, I found that [this](https://github.com/duplicati/duplicati/blob/c3136e31dfcf3c0857ed7f5eecf6cfb362d56df7/Duplicati/Library/Main/Operation/Backup/BackendUploader.cs#L122) `Task` would fault, the [code](https://github.com/duplicati/duplicati/blob/c3136e31dfcf3c0857ed7f5eecf6cfb362d56df7/Duplicati/Library/Main/Operation/Backup/BackendUploader.cs#L127) in the `finally` block would execute, and then [this call](https://github.com/duplicati/duplicati/blob/v2.0.4.15-2.0.4.15_canary_2019-02-06/Duplicati/Library/Main/Operation/BackupHandler.cs#L359) to `Task.WhenAny` would cause the exception to be unobserved.



Could it be that something about our test sources (size, number of volumes, etc.) is causing a slightly different code path to be taken?

**Comment by seantempleton** [comment]
@warwickmm I'm actually going to have to revise my statement. I forgot that my test backup had number-of-retries set to 0. There's actually a current bug where if you set retries to 0 then it doesn't even try to upload any files due to the code using < instead of <=. `for (var i = 0; i < m_options.NumberOfRetries; i++)`
Using the current code, 0 retries caused it to always look like it succeeded for me. When I set retries to 1 sometimes it would stop early and show an error and sometimes it would look like it succeeded.

At first I could only get my changes to stop and show an error when backing up about 300 files for a total of 300MB no matter how many retries I had it set to. Then I tried to backup a single 80KB file and it "succeeded"! Also with any number of retries.

I think you're right about the issue being the `Task.WhenAny` in the BackupHandler. The BackendUploader sets the result for the `flushReq` in the BackupHandler first before it exits itself. So the `WhenAny` will exit with the `flushReq` task and not call await on the `uploader`. That's my theory anyway, without actually debugging it.

**Comment by seantempleton** [comment]
@warwickmm are you planning on investigating this? If not I can give it a try.

**Comment by warwickmm** [comment]
If you can help with this, that would be fantastic.  I have tried several things and failed:
1. I can get the error to be raised, but it then appears that a backup entry is recorded in the database (a backup version appears in the restore list, but that backup doesn't really exist).
2. I then made some changes to avoid the errant entry in the database, but the next backup run throws a "A task has been cancelled" error.

So basically, I have not been successful yet.  Any help would be greatly appreciated!

**Comment by seantempleton** [comment]
The restore list looks like it comes from the `Fileset` table. A row is put in the table before any work is done when a backup is started. When the `Put` operation fails, even when all of them fail, that row is not removed. If you run a database repair then the backup version disappears from the UI and the `Fileset` row in the database is removed. All the other information, blocks, files, etc. are removed as well.

My guess is that when any `Put` operation fails that the database needs to be cleaned up once the backup has stopped. I don't know all of what the repair database option does. I would assume that to be too heavy handed in this case. We could delete all information about files, blocks, etc. from a specific dblock file when it fails to upload.

**Comment by warwickmm** [comment]
I'm not familiar with how the database transactions are currently handled, but I would think that the relevant transactions should be rolled back in event of a `Put` error.

**Comment by seantempleton** [comment]
The transactions are all committed by the time files begin uploading. If you put a breakpoint in the `BackendUploader` where it gets the first request, then when you hit the breakpoint and open the sqlite database you can see it has all the information in the tables already. The Block table, File, Fileset, BlockListHash, etc.

For instance the row in the `Fileset` table is inserted [in the BackupHandler](https://github.com/duplicati/duplicati/blob/master/Duplicati/Library/Main/Operation/BackupHandler.cs#L475) right before the call to `RunMainOperation`, which starts the process of scanning files, filtering files, calculating hashes, etc. And the transaction is committed so there's no possibility of rollback.

**Comment by warwickmm** [comment]
@seantempleton, I have a proposed fix in pull request #3697.  If you have a chance, can you take a look?

**Comment by warwickmm** [comment]
After some more testing, here's what I've discovered.

As of revision e885f214b0595b63a0b4354e3a6442808a5e4e2a:

* For larger backups (those that satisfy [uploadsInProgress >= m_maxConcurrentUploads](https://github.com/duplicati/duplicati/blob/e885f214b0595b63a0b4354e3a6442808a5e4e2a/Duplicati/Library/Main/Operation/Backup/BackendUploader.cs#L187)), failed Puts do result in an error (this appears to have been due to changes in pull request #3684 (thanks @seantempleton)).  The UI shows "Last successful backup: Never".  However, when attempting to restore, a backup version (with files) appears to exist, but restores will fail.

  The next successful backup runs without error.  However, we encounter a warning
  
  ```
  Expected there to be a temporary fileset for synthetic filelist (1, duplicati-20190407T022849Z.dlist.zip.aes), but none was found?
  ```
  The UI shows a nonzero backup size, but 0 backup versions and no versions appear in the resore list. The backend contains `dblock` and `dindex` files, but no `dlist` files.

  A subsequent successful backup then uploads the `dlist` file, a backup version appears in the UI and restore list, and restores work without issues.

* For smaller backups (those that do not satisfy [uploadsInProgress >= m_maxConcurrentUploads](https://github.com/duplicati/duplicati/blob/e885f214b0595b63a0b4354e3a6442808a5e4e2a/Duplicati/Library/Main/Operation/Backup/BackendUploader.cs#L187)), failed Puts **do not result in an error**.  The UI shows "0 bytes / 0 versions", and no versions appear in the restore list.

  The next successful backup runs without error, a backup version appears in the UI and restores list, and restores work without issues.


I have a fix (pull request #3720) that will make the behavior consistent for smaller backups.  Failed Puts result in an error and the UI shows "Last successful backup: Never".  However, when attempting to restore, a backup version (with files) appears to exist, but restores will fail.

The next successful backup runs without error.  However, we encounter a warning
```
Expected there to be a temporary fileset for synthetic filelist (1, duplicati-20190407T022849Z.dlist.zip.aes), but none was found?
```
The UI shows a nonzero backup size with 1 version, and restores work without issues.

I think the problem with the misleading restore version appearing when the Puts fail is a separate issue that we'll have to look into later.

**Comment by ts678** [comment]
> Expected there to be a temporary fileset for synthetic filelist (1, duplicati-20190407T022849Z.dlist.zip.aes), but none was found?

My theory on why this happens is posted in several places such as:

[Warning Message after Backup](https://forum.duplicati.com/t/warning-message-after-backup/965/7?u=ts678)

[Expected there to be a temporary fileset for synthetic filelist #2506](https://github.com/duplicati/duplicati/issues/2506#issuecomment-463622742)

Thanks for pushing on the problems of failed puts. Those are bound to happen sooner or later on networks.
